plugins/scenario/hu.bme.mit.gamma.scenario.language/src/hu/bme/mit/gamma/scenario/language/validation/ScenarioLanguageValidator.xtend

```diff
@@ -0,0 +1,393 @@
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * SPDX-License-Identifier: EPL-1.0
+ ********************************************************************************/
+package hu.bme.mit.gamma.scenario.language.validation
+
+import hu.bme.mit.gamma.expression.model.Expression
+import hu.bme.mit.gamma.expression.model.ExpressionModelPackage
+import hu.bme.mit.gamma.expression.util.ExpressionEvaluator
+import hu.bme.mit.gamma.scenario.language.util.StatechartLanguageUtil
+import hu.bme.mit.gamma.scenario.model.Chart
+import hu.bme.mit.gamma.scenario.model.CombinedFragment
+import hu.bme.mit.gamma.scenario.model.Delay
+import hu.bme.mit.gamma.scenario.model.Interaction
+import hu.bme.mit.gamma.scenario.model.InteractionDirection
+import hu.bme.mit.gamma.scenario.model.InteractionFragment
+import hu.bme.mit.gamma.scenario.model.LoopCombinedFragment
+import hu.bme.mit.gamma.scenario.model.ModalInteraction
+import hu.bme.mit.gamma.scenario.model.ModalInteractionSet
+import hu.bme.mit.gamma.scenario.model.ModalityType
+import hu.bme.mit.gamma.scenario.model.ParallelCombinedFragment
+import hu.bme.mit.gamma.scenario.model.Reset
+import hu.bme.mit.gamma.scenario.model.ScenarioDeclaration
+import hu.bme.mit.gamma.scenario.model.ScenarioDefinition
+import hu.bme.mit.gamma.scenario.model.ScenarioModelPackage
+import hu.bme.mit.gamma.scenario.model.Signal
+import hu.bme.mit.gamma.statechart.composite.SynchronousComponent
+import hu.bme.mit.gamma.statechart.interface_.EventDirection
+import hu.bme.mit.gamma.statechart.interface_.RealizationMode
+import java.util.List
+import java.util.Map
+import java.util.Set
+import java.util.concurrent.ConcurrentHashMap
+import java.util.stream.Collectors
+import org.eclipse.emf.common.util.EList
+import org.eclipse.emf.ecore.EObject
+import org.eclipse.emf.ecore.EStructuralFeature
+import org.eclipse.xtend.lib.annotations.Data
+import org.eclipse.xtext.EcoreUtil2
+import org.eclipse.xtext.validation.Check
+
+import static hu.bme.mit.gamma.scenario.language.util.EcoreUtilWrapper.*
+
+import static extension hu.bme.mit.gamma.scenario.util.CollectionUtil.myGet
+import static extension hu.bme.mit.gamma.scenario.util.CollectionUtil.myRemove
+import hu.bme.mit.gamma.scenario.model.NegatedModalInteraction
+import hu.bme.mit.gamma.scenario.model.StrictAnnotation
+import hu.bme.mit.gamma.scenario.model.PermissiveAnnotation
+import hu.bme.mit.gamma.scenario.model.NegStrictAnnotation
+import hu.bme.mit.gamma.scenario.model.NegPermissiveAnnotation
+
+/**
+ * This class contains custom validation rules. 
+ * 
+ * See https://www.eclipse.org/Xtext/documentation/303_runtime_concepts.html#validation
+ */
+class ScenarioLanguageValidator extends AbstractScenarioLanguageValidator {
```

Comment:
We should migrate the error message handling to the infrastructure of `ExpressionLanguageValidator`.